crates/ra_syntax/src/grammar/expressions.rs

```diff
@@ -246,19 +246,33 @@ fn lhs(p: &mut Parser, r: Restrictions) -> Option<CompletedMarker> {
             if p.at_ts(EXPR_FIRST) {
                 expr_bp(p, r, 2);
             }
-            return Some(m.complete(p, RANGE_EXPR));
+            return (Some(m.complete(p, RANGE_EXPR)), None);
         }
         _ => {
-            let lhs = atom::atom_expr(p, r)?;
-            return Some(postfix_expr(p, r, lhs));
+            let (lhs_marker, macro_block_like) = atom::atom_expr(p, r);
+
+            if let Some(lhs_marker) = lhs_marker {
+                return (
+                    Some(postfix_expr(p, r, lhs_marker, macro_block_like)),
+                    macro_block_like,
+                );
+            } else {
+                return (None, None);
+            }
         }
     };
     expr_bp(p, r, 255);
-    Some(m.complete(p, kind))
+    (Some(m.complete(p, kind)), None)
 }
 
-fn postfix_expr(p: &mut Parser, r: Restrictions, mut lhs: CompletedMarker) -> CompletedMarker {
-    let mut allow_calls = !r.prefer_stmt || !is_block(lhs.kind());
+fn postfix_expr(
+    p: &mut Parser,
+    r: Restrictions,
+    mut lhs: CompletedMarker,
+    macro_block_like: Option<BlockLike>,
```

Comment:
Similarly, we don't need `Option` here I guess?

Alternatively, I think we can just pass-in `allow_calls` directly, *instead* of both `r` and `block_like`.